src/landscape.cpp

```diff
@@ -992,29 +992,26 @@ static void CreateDesertOrRainForest(uint desert_tropic_line)
  */
 static bool FindSpring(TileIndex tile, void *)
 {
-	int reference_height;
-	if (!IsTileFlat(tile, &reference_height) || IsWaterTile(tile)) return false;
-
 	/* In the tropics rivers start in the rainforest. */
 	if (_settings_game.game_creation.landscape == LT_TROPIC && GetTropicZone(tile) != TROPICZONE_RAINFOREST) return false;
 
+	int reference_height;
+	if (IsWaterTile(tile) || !IsTileFlat(tile, &reference_height)) return false;
+
 	/* Are there enough higher tiles to warrant a 'spring'? */
 	uint num = 0;
-	for (int dx = -1; dx <= 1; dx++) {
-		for (int dy = -1; dy <= 1; dy++) {
-			TileIndex t = TileAddWrap(tile, dx, dy);
-			if (t != INVALID_TILE && GetTileMaxZ(t) > reference_height) num++;
-		}
+	for (Direction d = DIR_BEGIN; d != DIR_END; d++) {
+		TileIndex t = AddTileIndexDiffCWrap(tile, TileIndexDiffCByDir(d));
+		if (IsValidTile(t) && GetTileMaxZ(t) > reference_height) num++;
+		if (num == 4) break;
 	}
```

Comment:
Why not use `TileArea(tile, 1, 1).Expand(1)` to iterate over?
If it works for the 33x33 area below, why wouldn't it work here?

Reply:
`TileArea` approach would include the tile in the center, which is already proven to be flat in the check above. Its height is known and matches the `reference_height`. `GetTileMaxZ` of this tile would fail the `> reference_height`. It's safe to exclude it.